feat: Add Ollama module

frankhaugen opened the pull request:
## What does this PR do?

Added a new Module project that uses OLLAMA docker images to give local LLM capabilities. This only supports CPU workload, but adding the env. variables to activate GPU if the prerequisites are there, can be done and so CUDA core based processing isn't precluded

## Why is it important?

This module is very powerful and can give a lot of value to those that are incorporating OLLAMA models in their projects

## Related issues

Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #1097

## How to test this PR

Run the test added

## Follow-ups

- Make more docs to guide the user to be able to use GPU might be a future thing, but right now, no need.

frankhaugen committed:
Initial create the module from template

frankhaugen committed:
Added test project for Ollama module

frankhaugen committed:
Correct any "wrongness" for setting up project and test

frankhaugen committed:
Added the nitty-gritty of the module

frankhaugen committed:
Cleanup

frankhaugen committed:
Merge branch 'develop' into frank/ollama-module

HofmeisterAn commented on src/Testcontainers.Ollama/OllamaContainer.cs:
```diff
@@ -0,0 +1,52 @@
+            }));
+            
+            await Task.CompletedTask;
+        }
```
Does it make sense to pass the model name as an argument to the member? `Run(string modelName, CancellationToken ct = default)`? In addition, please do not block the thread; simply return `ExecAsync(["ollama", "run", modelName], ct);`.

frankhaugen replied:
Agreed and resolved about the return. I added two methods one taking only the ct and one taking name and ct. Have a look if you agree

HofmeisterAn commented on src/Testcontainers.Ollama/OllamaConfiguration.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <summary>
+        /// Default volume name.
+        /// </summary>
+        public const string DefaultVolumeName = "ollama-volume";
```
Can you please move the default configurations to the builder, like all the other modules do?

frankhaugen replied:
Done

HofmeisterAn commented on src/Testcontainers.Ollama/OllamaContainer.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            if (State!= TestcontainersStates.Created && State != TestcontainersStates.Running)                {
+                throw new InvalidOperationException("Cannot start a container that has not been created.");
+            }
```
We use `ThrowIfResourceNotFound();` to make sure the resources exists and is "running".

frankhaugen replied:
Done

HofmeisterAn commented on src/Testcontainers.Ollama/OllamaConfiguration.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <summary>
+        /// Default image name.
+        /// </summary>
+        public const string ImageName = "ollama/ollama:latest";
```
Please pin the image version.

frankhaugen replied:
Oops, that was an oversight

HofmeisterAn commented on tests/Testcontainers.Ollama.Tests/OllamaContainerTests.cs:
```diff
@@ -0,0 +1,49 @@
+
+    public async Task InitializeAsync()
+    {
+      TestcontainersSettings.Logger = new TestOutputLogger(nameof(OllamaContainerTests), _testOutputHelper);
```
Can we please keep the default logger configuration, etc. I am aware of the issues around xUnit.net's `ITestOutputHelper` and console logging, but I would prefer to keep the modules and tests consistent. This helps in applying future refactoring, etc.

frankhaugen replied:
I was a bit careless, and shouldn't have included it, its very much a "I like this to as a debug tool while working" so, removed

HofmeisterAn reviewed (changes requested):
Thanks for the PR. I had a quick look and added some minor suggestions and improvements.

frankhaugen committed:
#1097 Update Ollama configuration and remove test helpers

This commit updates the Ollama configuration to allow more customization options and removes unnecessary test helpers. The OllamaConfiguration class was refactored to provide more configurable parameters such as the VolumePath and VolumeName. Additionally, the TestOutputHelperExtensions and TestOutputLogger classes were deleted as they were not providing any significant value.

frankhaugen committed:
Merge remote-tracking branch 'origin/frank/ollama-module' into frank/ollama-module

frankhaugen committed:
Merge branch 'develop' into frank/ollama-module

frankhaugen committed:
Merge branch 'develop' into frank/ollama-module

frankhaugen commented:
@HofmeisterAn  I'm updating the branch with my current state as the tests were green on local, but I think it needs another pass. No Rush, I'm busy with my day-job so, no need for you to rush 😄

HofmeisterAn committed:
Merge branch 'develop' into frank/ollama-module

eddumelendez commented on src/Testcontainers.Ollama/OllamaBuilder.cs:
```diff
@@ -0,0 +1,121 @@
+        /// <summary>
+        /// The default model name for the OllamaBuilder.
+        /// </summary>
+        public const string DefaultModelName = OllamaModels.Llama2;
```
Models are just like images and should be provided as a string rather than having an specific type. It will also reduce the amount of PRs just to keep it up-to-date

eddumelendez commented on src/Testcontainers.Ollama/OllamaBuilder.cs:
```diff
@@ -0,0 +1,121 @@
+        /// <summary>
+        /// Default volume path.
+        /// </summary>
+        public const string DefaultVolumePath = "/root/.ollama";
```
Volume mounting should be optional rather than the default, at least for now. We can align along with Java and Go implementation.

eddumelendez reviewed (commented):
Thanks for your contribution. Testcontainers for Java and Go has recently release a new Ollama module, which allows to detect and enable GPU and allow to commit the container's state and create a new Image. It would be nice to align the .NET implementation along with the others. See [java implementation](https://github.com/testcontainers/testcontainers-java/pull/8369).

duranserkan commented:
Is there any update on this?

IRusio commented:
any update on that?

HofmeisterAn committed:
Merge remote-tracking branch 'origin/develop' into fork/frankhaugen/frank/ollama-module

HofmeisterAn committed:
chore: Address merge conflicts (update to develop)

HofmeisterAn reviewed (commented):
I updated the PR to develop. I'll take a look at it sometime next week and address the necessary parts.

HofmeisterAn committed:
chore: Update module

HofmeisterAn committed:
chore: Add test

HofmeisterAn reviewed (approved):
I updated the module implementation and aligned it with the other Testcontainers module implementations. I also added a text example.

> Thanks for your contribution. Testcontainers for Java and Go has recently release a new Ollama module, which allows to detect and enable GPU and allow to commit the container's state and create a new Image. It would be nice to align the .NET implementation along with the others. See java implementation.

I didn't include detecting and setting the GPU configuration, as it's not that simple in .NET (but it's something we can address in the future). I'm also not sure if committing an image is supported in Docker.DotNet. I need to double-check that. It might be something we need to implement first. If someone wants to pick it up, that would be great, but we need to discuss how to address this in an issue first.

```csharp
// How to set the device requests using the container builder API:
.WithCreateParameterModifier(parameterModifier =>
    parameterModifier.HostConfig.DeviceRequests
        = new List<DeviceRequest> { new DeviceRequest { /* TODO: Set properties. */ } })
```

kiview commented:
I also like to reference the `useLocal` feature in the [tc-go implementation](https://golang.testcontainers.org/modules/ollama/#use-local). Pragmatically, it is the best way to get a good DX on macOS.
